File: Airwallex/Airwallex/Sources/Models/AWXPaymentMethodOptions.m
//
//  AWXPaymentMethodOptions.m
//  Airwallex
//
//  Created by Victor Zhu on 2020/1/16.
//  Copyright © 2020 Airwallex. All rights reserved.
//

#import "AWXPaymentMethodOptions.h"

@implementation AWXThreeDs

- (NSDictionary *)encodeToJSON
{
    NSMutableDictionary *threeDs = [NSMutableDictionary dictionary];
    if (self.paRes) {
        threeDs[@"pa_res"] = self.paRes;
    }
    if (self.returnURL) {
        threeDs[@"return_url"] = self.returnURL;
    }
    if (self.attemptId) {
        threeDs[@"attempt_id"] = self.attemptId;
    }
    if (self.deviceDataCollectionRes) {
        threeDs[@"device_data_collection_res"] = self.deviceDataCollectionRes;
    }
    if (self.dsTransactionId) {
        threeDs[@"ds_transaction_id"] = self.dsTransactionId;
    }
    return threeDs;
}

@end

@implementation AWXCardOptions

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.autoCapture = YES;
    }
    return self;
}

- (NSDictionary *)encodeToJSON
{
    NSMutableDictionary *options = [NSMutableDictionary dictionary];
    if (self.threeDs) {
        options[@"three_ds"] = [self.threeDs encodeToJSON];
    }
    options[@"auto_capture"] = @(self.autoCapture);
    return options;
}

@end

@implementation AWXPaymentMethodOptions

- (NSDictionary *)encodeToJSON
{
    NSMutableDictionary *options = [NSMutableDictionary dictionary];
    if (self.cardOptions) {
        options[@"card"] = [self.cardOptions encodeToJSON];
    }
    return options;
}

@end

File: Airwallex/Core/Sources/Internal/Models/AWXPaymentMethodOptions.swift
//
//  AWXPaymentMethodOptions.swift
//  Core
//
//  Created by Tony He (CTR) on 2024/7/22.
//  Copyright © 2024 Airwallex. All rights reserved.
//

import Foundation

/// `AWXPaymentMethodOptions` includes the info of payment consent.
@objcMembers
@objc
public class AWXPaymentMethodOptions: NSObject, Codable {
    /**
     The options for card.
     */
    public let cardOptions: AWXCardOptions?

    enum CodingKeys: String, CodingKey {
        case cardOptions = "card"
    }

    public init(cardOptions: AWXCardOptions?) {
        self.cardOptions = cardOptions
    }

    public static func decodeFromJSON(_ dic: [String: Any]) -> AWXPaymentMethodOptions {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: dic, options: [])
            let decoder = JSONDecoder()
            let result = try decoder.decode(AWXPaymentMethodOptions.self, from: jsonData)

            return result
        } catch {
            return AWXPaymentMethodOptions(cardOptions: nil)
        }
    }

    public func encodeToJSON() -> [String: Any] {
        return toDictionary() ?? [String: Any]()
    }
}
